bpo-21261: IDLE - add completion of dict sting keys

taleinat opened the pull request:
This is a new implementation, created from scratch. I intentionally didn't base this on PR GH-1511 after reviewing it thoroughly.

Note that completion is intentionally limited to keys of type str and bytes only.

Additional changes made in this PR:
* The Escape key now closes the completion window. [already done separately]
* Completing a string or bytes dict key adds a closing square bracket.
* Some of the auto-completion tests have been changed a bit to run without a GUI.
* Significant refactoring and cleanup of some existing tests, with better use of mocking.
* The fake Text widget used for tests has had its index handling greatly improved.
* Added a very functional fake Listbox widget class.

<!-- issue-number: [bpo-21261](https://bugs.python.org/issue21261) -->
https://bugs.python.org/issue21261
<!-- /issue-number -->

taleinat committed:
prefix preservation for str and bytes

taleinat committed:
refactor back to original interface of fetch_completions()

taleinat committed:
refactor existing tests and improve mock_tk.Text._decode()

taleinat committed:
add tests

taleinat committed:
add a NEWS entry

taleinat committed:
code cleanup and added comments

aeros commented on Lib/idlelib/autocomplete.py:
```diff
@@ -90,6 +92,126 @@ def _delayed_open_completions(self, args):
+            12: '\\f',
+            13: '\\r',
+        })
+        high_byte_escape_re = re.compile(r'[\x80-\xff]')
```
Based on the [travis error log](https://travis-ci.org/python/cpython/jobs/569053155), it looks like this line is likely the culprit (missing empty line before function):
```suggestion
        high_byte_escape_re = re.compile(r'[\x80-\xff]')
        
```

aeros commented on Misc/NEWS.d/next/IDLE/2019-08-07-23-27-54.bpo-21261.sqNQCD.rst:
```diff
@@ -0,0 +1 @@
+IDLE's shell auto-completion completes dict keys of type str and bytes.
```
The main suggestion here is adjusting "auto-completion completes" -> "auto-completion works for" and "type" -> "types". 

The inline markup for str and bytes is an extra addition, sphinx converts this into a link to their stdtypes sections. Respectively: [str](https://docs.python.org/3.9/library/stdtypes.html#str) and [bytes](https://docs.python.org/3.9/library/stdtypes.html#bytes). For an example of this being used in the docs, see [common sequence operators](https://docs.python.org/3.9/library/stdtypes.html#common-sequence-operations), in the section for the first note. If you don't want the link, it could be instead replaced with ``str`` and ``bytes``.

```suggestion
IDLE's shell auto-completion works for dict keys of types :class:`str` and :class:`bytes`.
```

taleinat replied:
Great suggestions, thanks!

aeros reviewed (approved):
Tested using the IDLE shell from the PR branch. OS: Arch Linux 5.2.3. 

Test for str
Before tab:
![image](https://user-images.githubusercontent.com/44193521/62674517-bc693d00-b970-11e9-9346-ef25b8873a1f.png)
After tab:
![image](https://user-images.githubusercontent.com/44193521/62674533-cab75900-b970-11e9-916a-6d195897b2de.png)

Test for bytes
Before tab:
![image](https://user-images.githubusercontent.com/44193521/62674613-271a7880-b971-11e9-807b-f390770ac55d.png)
After tab:
![image](https://user-images.githubusercontent.com/44193521/62674623-2e418680-b971-11e9-97c2-7e544ae3ca92.png)


Looks like it's working as intended, nicely done! I also tested to ensure that ``excessively_long_key`` and didn't auto-complete when using ``d[b'e`` and the same for ``d['t`` -> ``d['thanks taleinat'``. I intentionally used a space instead of an underscore for the second one (even if it goes against naming convention) to ensure there weren't any issues there. LGTM.

aeros commented on Lib/idlelib/autocomplete.py:
```diff
@@ -90,6 +92,126 @@ def _delayed_open_completions(self, args):
+        bytes_comps = (c for c in comp_list if type(c) == bytes)
+
+        # Create reprs for the str objects.
+        str_repr_template = f"{prefix}{quote_type}{{}}{quote_type}"
```
Minor QoL suggestion:

In any scenario I can imagine for the dict key auto-complete, the user would end up typing `]` immediately after using `tab`. As a result, it might be beneficial to add it to the end of the auto-completion.

The best place to add this would probably be in the string template:
```suggestion
        str_repr_template = f"{prefix}{quote_type}{{}}{quote_type}]"
```

aeros replied:
Tested this suggestion locally, no issues as far as I can tell.

Before tab:
![image](https://user-images.githubusercontent.com/44193521/62675809-7fa04480-b976-11e9-90fd-9f01bffaf26c.png)
After tab:
![image](https://user-images.githubusercontent.com/44193521/62675824-8c249d00-b976-11e9-891b-682c52d79d7d.png)

taleinat replied:
This is a great idea and I've just implemented it, albeit using a more robust and complete approach.

aeros replied:
Awesome! I definitely agree with the usage of a variable instead of appending a constant, that allows far more extensibility. I hadn't thought of using anything besides `]`, but that could definitely be useful.

Edit: I also would not have though of adding this to the `keypress_event()` function instead of just appending as a string to the output. Good idea.

aeros commented on Lib/idlelib/autocomplete.py:
```diff
@@ -90,6 +92,126 @@ def _delayed_open_completions(self, args):
+    _dict_str_key_prefix_re = re.compile(r'\s*(br|rb|fr|rf|b|f|r|u|)', re.I)
+
+    @classmethod
+    def is_completing_dict_key(cls, hp, curline):
```
Another minor suggestion. While reading over the changes, the purpose of the *hp* argument in `is_completing_dict_key()` was not clear until I saw line 228. The variable name `hp` isn't as much of an issue on 228 where it's defined, but I think it would make the purpose of the argument in `is_completing_dict_key()` much more clear if it was instead *parser* or *hparser*:

```suggestion
    def is_completing_dict_key(cls, hparser, curline):
```
With this change, anyone reading over the function definition could much more easily understand the purpose of the argument without having to understand the exact implementation details. 

I prefer *hparser*, to avoid any potential confusion with argparse, since *parser* is frequently used as a variable name when creating an instance of the ArgumentParser class.

taleinat replied:
A fine suggestion, I've just done this.

taleinat committed:
whitespace fix

taleinat committed:
Escape key closes the completion window

taleinat committed:
a bit of code cleanup

taleinat committed:
many major improvements

taleinat committed:
implement try-open for dict keys, and fix handling of ']'

taleinat committed:
fix whitespace

taleinat committed:
improve NEWS entry wording and formatting

taleinat committed:
_is_completing_dict_key args: rename 'hp' and remove unused 'curline'

aeros commented on Lib/idlelib/autocomplete_w.py:
```diff
@@ -153,6 +173,39 @@ def _selection_changed(self):
+            i -= 2
+        return start[i - 1] != '\\'
+
     def show_window(self, comp_lists, index, complete, mode, userWantsWin):
```
Could we potentially change the parameter `userWantsWin` -> `user_wants_win` or `wants_win`? This already existed prior to this change, but it looks a bit odd to have one parameter within the function use the typical underscore naming convention and another using camelCase.

```suggestion
     def show_window(self, comp_lists, index, complete, mode, wants_win):
```

terryjreedy replied:
Yes, we have be gradually PEP8-ifying idlelib as modules are otherwise reviewed and edited *and tests added*.  And shorter is fine here.

taleinat replied:
This argument is actually no longer used for anything as far as I can tell. I double-check and then remove it.

taleinat replied:
Removed in latest commit.

aeros reviewed (approved):
Tested the dict key auto-completion again after the most recent changes:

Before tab:
![image](https://user-images.githubusercontent.com/44193521/62750306-b08f8080-ba2d-11e9-8770-9cc37b70ba2b.png)
After tab:
![image](https://user-images.githubusercontent.com/44193521/62750317-bd13d900-ba2d-11e9-9d33-24bdb7afea9e.png)
Looks like it's worked as intended. Nicely done.

Let me know if there's any more in-depth testing you'd like for me to do.

taleinat committed:
remove unused "user wants win" variables and function args

aeros commented on Lib/idlelib/autocomplete_w.py:
```diff
@@ -24,6 +29,13 @@
 
+
+_quote_re = re.compile(r"""["']""")
+def _find_first_quote(s):
```
```suggestion
def _find_first_quote(line):
```

Edit: Oops, this should be line, not quote.

taleinat replied:
Thanks for the suggestion! I've done this, and otherwise improved the function e.g. adding a doc-string.

aeros replied:
Thanks! Doc-strings are always a good idea to include.

aeros commented on Lib/idlelib/autocomplete_w.py:
```diff
@@ -24,6 +29,13 @@
+
+_quote_re = re.compile(r"""["']""")
+def _find_first_quote(s):
+    m = _quote_re.search(s)
```
```suggestion
    m = _quote_re.search(line)
```

aeros reviewed (commented):
This is a similar recommendation to the one in first PR review I had made, changing a single letter parameter name to something slightly more descriptive. It might not be needed in this case since it's a brief underscore function, but I figured it wouldn't hurt to ask. I don't think `m` as a variable name is an issue, since it's used frequently enough for regex matches and not a parameter.

Edit: The variable suggestion should be `line`, not quote.

taleinat committed:
clean up code of _find_first_quote()

taleinat committed:
remove unnecessary "from tkinter import *"

terryjreedy reviewed (changes requested):
Initial use test:
`>>> globals()[`
Wow.  The popup seems so natural I had to remind myself that it is new.  I definitely want this in 3.8.0b4, due  Monday, 2019-08-26.

But... given the current completion design, calling a function for an auto popup is a bug.  Dict completion should act the same as attribute completion (and calltips).
`>>> globals().`  # No attribute completion popup unless explicitly request (force, control-space).
`>>> g = globals(); g.`  # Attribute completion pops up, starting with 'clear'.

Bug 2.  If I close the box (Esc, immediate Enter (see below), click outside the box) and enter a 2nd '[', the 1st '[' is ignored and a box pop up.  Same if close and enter 3rd '[', and so on
`>>> g[[`  # Completion box for g pops up.  But `get_entity('g[')` (or for 'g[[', etc) should fail.

I would have preferred at least 2 separate PRs, the first being for changing existing tests and improving mock_tk.  But separate PRs would add calendar time, and I realize that you probably did not know what you wanted to add before working on new tests.  I will, however, review those separately.  It helps a lot that you make multiple commits and don't squash them.

Though a separate issue, closing with Escape is great, and part of making all popups close with Escape.  It opens the opportunity to reassess how completion boxes open and possibly no longer using Enter to close without completing.  But let's leave that for another issue.  I prefer no more feature creep on this PR ;-).  This addition should be mentioned on the blurb.

idle.rst (and help.html with it) need revision.  I can do that.

taleinat committed:
fix opening of dict key completion for expressions with func calls

aeros commented on Lib/idlelib/autocomplete.py:
```diff
@@ -203,6 +396,14 @@ def fetch_completions(self, what, mode):
+                    keys = entity.keys() if isinstance(entity, dict) else []
+                except:
+                    return [], []
+                smalll = bigl = [k for k in keys if type(k) in (bytes, str)]
```
> Bug 3. The keys completion box should only have the dict keys, as with attributes, but not the global list.

@terryjreedy Could we fix this by adding another conditional filter to the list comprehension?

```suggestion
                smalll = bigl = [k for k in keys if type(k) in (bytes, str) and k not in globals()]
```

taleinat replied:
@aeros167, this was done intentionally (see details in separate comment.)

terryjreedy replied:
I don't see from the snippet given how the 230+ globals are there.  But they should never be in keys.

aeros replied:
>I don't see from the snippet given how the 230+ globals are there. But they should never be in keys.

@terryjreedy Hmm, this suggestion might be in the wrong location then. This PR was the first time I've read through idlelib/autocomplete.py. Based on my understanding, this was list of potential completions for dict keys, so filtering out any of the keys which were located in ``globals()`` would address bug 3.

@taleinat I made another suggestion in a comment below which performs filtering from ``globals()`` only when the completion is started after an opening quote.

aeros commented on Lib/idlelib/autocomplete.py:
```diff
@@ -90,14 +93,177 @@ def _delayed_open_completions(self, args):
+                quote_type_candidate = post_prefix[0]
+                if quote_type_candidate not in "'\"":
+                    pass  # use the default quote type
+                elif post_prefix[1:2] == quote_type_candidate:
```
> When completion is invoked after an opening quote, it is clear that a string completion is desired, and perhaps then only string/bytes completions should be offered.

@taleinat Could this be done by filtering ``comp_list`` after ``quote_type_candidate`` is determined to be either ``'`` or ``"``?
```suggestion
                comp_list = [c for c in comp_list if c not in globals()]
                elif post_prefix[1:2] == quote_type_candidate:
```
Or is there a better location to do this?

terryjreedy commented on Lib/idlelib/autocomplete.py:
```diff
@@ -106,7 +272,17 @@ def open_completions(self, args):
         i = j = len(curline)
-        if hp.is_in_string() and (not mode or mode==FILES):
+
+        if mode in (None, DICTKEYS):
```
It looks like this suite should be in the block for `elif mode in (None, DICTKEYS):` (revised from  `elif mode == DICTKEYS:` ).

taleinat replied:
I'm not sure what you mean... I put this block at the top intentionally. Note that it is followed by `if` rather than `elif`, so that if the block doesn't decide that dict key completion should take place, the other completion modes are checked.

terryjreedy replied:
I did notice, but it is not yet obvious why the preliminary check in needed if it can be overriden.  But leave it for now.

terryjreedy commented on Lib/idlelib/autocomplete.py:
```diff
@@ -137,18 +313,35 @@ def open_completions(self, args):
+            if hp.is_in_code():
+                # Add variables in the global namespace to the comp lists.
+                comp_lists = (reprs, list(reprs))
+                extra_comp_lists = self.fetch_completions("", ATTRS)
```
This should be deleted.

taleinat replied:
Please see my comment above re. intentionally including global variables in the lists.

terryjreedy replied:
I will try to explain better in a post to the issue why I am currently opposed to this.

terryjreedy reviewed (changes requested):
A bare beginning of code review.  It is more complicated than I hoped.  I see the problem with strings (bytes) just being strings, while string (bytes) literals have all the complications.  Even without that, strings can be indefinitely long, so we have to truncate.  A later issue could be to allow a wider box if needed.
Replacing newlines in multiline strings with '\n' seems like a good idea also.

taleinat committed:
fix bug opening completions in the middle of a dict key string

taleinat committed:
escape non-BMP unicode characters in dict key string completions

taleinat committed:
refactor _dict_key_reprs() to avoid compiling functions on every call

taleinat committed:
update idle.rst

taleinat committed:
add a What's New entry

terryjreedy commented on Lib/idlelib/autocomplete.py:
```diff
@@ -183,10 +183,14 @@ def control_char_escape(match):
         def high_byte_sub(match):
             """escaping for bytes values above 127"""
             return f'\\x{ord(match.group()):x}'
+        non_bmp_re = re.compile(r'[\U00010000-\U0010ffff]')
```
As suggested by my email, this and the function or something like them, and probably the code that calls it, is needed many other places.  But I don't know yet where it should go.  Thanks for the start.

terryjreedy commented on Lib/idlelib/autocomplete.py:
```diff
@@ -183,10 +183,14 @@ def control_char_escape(match):
         def high_byte_sub(match):
```
I think non-ascii BMP chars should be left alone, as they currently are for tab-completion. 
```
>>> あいうえ = 3
>>> a<tab>
```
displays a box with `あいうえ` sorted at the end. `\\x3042\\x3044\\x3046\\x3048` would be useless to most people using hiragana.  I think we should optimize for people who sensibly limit string keys as well names to printable unicode.

taleinat replied:
The "high bytes" substitution is only used for bytes literals, and it is required for those literals to be valid. According to [the docs](https://docs.python.org/3/reference/lexical_analysis.html#string-and-bytes-literals):

>  They may only contain ASCII characters; bytes with a numeric value of 128 or greater must be expressed with escapes.

terryjreedy replied:
Got it.

taleinat committed:
use sys.maxunicode and add doc-strings and informative comments

taleinat committed:
be extra-safe and don't crash on import if sys.maxunicode < 0x10000

decaz commented on Lib/idlelib/autocomplete.py:
```diff
@@ -4,29 +4,32 @@
 import string
 import sys
 
 # Two types of completions; defined here for autocomplete_w import below.
```
*Three* types (or modes?) of completions; ...

taleinat replied:
Yes, thanks!
```suggestion
# Three types of completions; defined here for autocomplete_w import below.
```

taleinat committed:
update comment

taleinat committed:
Merge branch 'master' into idle-dict-key-completion

taleinat committed:
fix completing after a bracket with no available dict key completions

Also added tests for this case, as well as for typing an opening
bracket in a string, which should not trigger dict key completion.

taleinat committed:
add a couple more test cases

taleinat committed:
Merge branch 'master' into idle-dict-key-completion

# Conflicts:
#	Doc/whatsnew/3.8.rst

taleinat committed:
revert outdated additions to "What's New" docs

taleinat committed:
Merge branch 'master' into idle-dict-key-completion

# Conflicts:
#	Doc/library/idle.rst
#	Lib/idlelib/autocomplete.py
#	Lib/idlelib/autocomplete_w.py
#	Lib/idlelib/idle_test/test_autocomplete.py
